Upstream-Status: Pending
Signed-off-by: Armin Kuster <akuster808@gmail.com>

Index: src/tpm_library.c
===================================================================
--- git.orig/src/tpm_library.c
+++ git/src/tpm_library.c
@@ -405,8 +405,8 @@ int TPMLIB_LogPrintf(const char *format,
     }
 
     if (debug_prefix)
-        dprintf(debug_fd, debug_prefix);
-    dprintf(debug_fd, buffer);
+        dprintf(debug_fd, "%s" , debug_prefix);
+    dprintf(debug_fd, "%s" , buffer);
 
     return i;
 }
