Index: freeDiameter-8662db9f6105/tests/CMakeLists.txt
===================================================================
--- freeDiameter-8662db9f6105.orig/tests/CMakeLists.txt
+++ freeDiameter-8662db9f6105/tests/CMakeLists.txt
@@ -107,7 +107,6 @@ ENDIF(BUILD_APP_ACCT OR ALL_EXTENSIONS)
 FOREACH( TEST ${TEST_LIST} )
    ADD_EXECUTABLE(${TEST} ${TEST}.c tests.h ${${TEST}_ADDITIONAL})
    TARGET_LINK_LIBRARIES(${TEST} libfdproto libfdcore ${GNUTLS_LIBRARIES} ${GCRYPT_LIBRARY} ${${TEST}_ADDITIONAL_LIB})
-   ADD_TEST(${TEST} ${EXECUTABLE_OUTPUT_PATH}/${TEST} ${TEST_ARGUMENTS})
 ENDFOREACH( TEST )
 
 
