Index: tpm-tools-1.3.8/src/tpm_mgmt/tpm_present.c
===================================================================
--- tpm-tools-1.3.8.orig/src/tpm_mgmt/tpm_present.c
+++ tpm-tools-1.3.8/src/tpm_mgmt/tpm_present.c
@@ -349,13 +349,13 @@ int main(int argc, char **argv)
 		}
 	} while (flags[++i].name);
 
-      out_success:
+out_success:
 	logSuccess(argv[0]);
 	iRc = 0;
-      out_close:
+out_close:
 	contextClose(hContext);
-      out:
-    if (szTpmPasswd && !isWellKnown)
-	shredPasswd( szTpmPasswd );
-	return iRc;
+out:
+	if (szTpmPasswd && !isWellKnown)
+		shredPasswd( szTpmPasswd );
+return iRc;
 }
