From 0fd50756af998fabc3393f67ff8b78a4855303d9 Mon Sep 17 00:00:00 2001
From: Andreas Cadhalpun <Andreas.Cadhalpun@googlemail.com>
Date: Mon, 4 Aug 2014 22:44:40 +0200
Subject: remove unnecessary/harmful flags from libclamav.pc

---
 libclamav.pc.in | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/libclamav.pc.in b/libclamav.pc.in
index 27db30e757ff..19ef1e4e0399 100644
--- a/libclamav.pc.in
+++ b/libclamav.pc.in
@@ -6,6 +6,6 @@ includedir=@includedir@
 Name: libclamav
 Description: A GPL virus scanner
 Version: @PACKAGE_VERSION@
-Libs: -L${libdir} -lclamav @LIBCLAMAV_LIBS@
-Cflags: -I${includedir} @CFLAGS@
+Libs: -L${libdir} -lclamav
+Cflags: -I${includedir}
 
