samhain: musl build fix c99

Upstream-Status: Submitted
Signed-off-by: Armin Kuster <akuster@mvista.com>

Index: samhain-4.2.0/src/dnmalloc.c
===================================================================
--- samhain-4.2.0.orig/src/dnmalloc.c
+++ samhain-4.2.0/src/dnmalloc.c
@@ -5524,7 +5524,7 @@ arc4_stir(void)
         struct {
                 struct timeval tv1;
                 struct timeval tv2;
-                u_int rnd[(128 - 2*sizeof(struct timeval)) / sizeof(u_int)];
+                unsigned char rnd[(128 - 2*sizeof(struct timeval)) / sizeof(unsigned char)];
         } rdat;
 #if !defined(__FreeBSD__) && !defined(__OpenBSD__) && !defined(__NetBSD__)
         ssize_t sz = 0;
