
Upstream-Status: Inappropriate [OE config]

Signed-off-by: Armin Kuster <akuster808@gmail.com>

Index: configure.ac
===================================================================
--- a/configure.ac
+++ b/configure.ac
@@ -349,21 +349,17 @@ CFLAGS="$CFLAGS -Wformat -Wformat-securi
 dnl We have to make sure libtpms is using the same crypto library
 dnl to avoid problems
 AC_MSG_CHECKING([the crypto library libtpms is using])
-dirs=$($CC $CFLAGS -Xlinker --verbose 2>/dev/null | \
-       sed -n '/SEARCH_DIR/p' | \
-       sed 's/SEARCH_DIR("=\?\(@<:@^"@:>@\+\)"); */\1\n/g')
-for dir in $dirs; do
-  if test -r $dir/libtpms.so; then
-    if test -n "`ldd $dir/libtpms.so | grep libcrypto.so`"; then
-      libtpms_cryptolib="openssl"
-      break
-    fi
-    if test -n "`ldd $dir/libtpms.so | grep libnss3.so`"; then
-      libtpms_cryptolib="freebl"
-      break
-    fi
+dir="$SEARCH_DIR"
+if test -r $dir/libtpms.so; then
+  if test -n "`ldd $dir/libtpms.so | grep libcrypto.so`"; then
+    libtpms_cryptolib="openssl"
+    break
   fi
-done
+  if test -n "`ldd $dir/libtpms.so | grep libnss3.so`"; then
+    libtpms_cryptolib="freebl"
+    break
+  fi
+fi
 
 if test -z "$libtpms_cryptolib"; then
   AC_MSG_ERROR([Could not determine libtpms crypto library.])
