From: Yulong Pei <Yulong.pei@windriver.com>
Date: Wed, 21 Jul 2010 22:33:43 +0800
Subject: [PATCH] change finding path of nss and nspr

Upstream-Status: Pending

Signed-off-by: Yulong Pei <Yulong.pei@windriver.com>
Signed-off-by: Mingli Yu <Mingli.Yu@windriver.com>
---
 configure.ac |   12 ++++++------
 1 files changed, 6 insertions(+), 6 deletions(-)

--- a/configure.ac
+++ b/configure.ac
@@ -618,7 +618,7 @@
 
     if test "z$with_nspr" != "z" ; then
 	NSPR_PREFIX="$with_nspr"
-	NSPR_CFLAGS="-I$with_nspr/include -I$with_nspr/include/nspr"
+	NSPR_CFLAGS="-I$with_nspr/usr/include -I$with_nspr/usr/include/nspr4"
 	if test "z$with_gnu_ld" = "zyes" ; then
 	    NSPR_LIBS="-Wl,-rpath-link -Wl,$with_nspr/lib -L$with_nspr/lib $NSPR_LIBS_LIST"
 	else
@@ -626,7 +626,7 @@
 	fi
 	NSPR_INCLUDES_FOUND="yes"
 	NSPR_LIBS_FOUND="yes"
-	NSPR_PRINIT_H="$with_nspr/include/prinit.h"
+	NSPR_PRINIT_H="$with_nspr/usr/include/nspr4/prinit.h"
     else
 	for dir in $ac_nss_inc_dir ; do
     	    if test -f $dir/nspr/prinit.h ; then
@@ -664,7 +664,7 @@
 	OLD_CPPFLAGS=$CPPFLAGS
 	CPPFLAGS="$NSPR_CFLAGS"
 	AC_EGREP_CPP(yes,[
-    	    #include <prinit.h>
+	    #include <nspr4/prinit.h>
             #if PR_VMAJOR >= 4
                yes
             #endif
@@ -689,7 +689,7 @@
     NSS_NSS_H=""
 
     if test "z$with_nss" != "z" ; then
-	NSS_CFLAGS="$NSS_CFLAGS -I$with_nss/include -I$with_nss/include/nss"
+	NSS_CFLAGS="$NSS_CFLAGS -I$with_nss/usr/include -I$with_nss/usr/include/nss3 -I$with_nspr/usr/include/nspr4"
 	if test "z$with_gnu_ld" = "zyes" ; then
 	    NSS_LIBS="$NSS_LIBS -Wl,-rpath-link -Wl,$with_nss/lib -L$with_nss/lib $NSS_LIBS_LIST"
         else
@@ -697,7 +697,7 @@
         fi
 	NSS_INCLUDES_FOUND="yes"
 	NSS_LIBS_FOUND="yes"
-	NSS_NSS_H="$with_nss/include/nss.h"
+	NSS_NSS_H="$with_nss/usr/include/nss3/nss.h"
     else
 	for dir in $ac_nss_inc_dir ; do
 	    if test -f $dir/nss/nss.h ; then
@@ -735,7 +735,7 @@
        OLD_CPPFLAGS=$CPPFLAGS
        CPPFLAGS="$NSS_CFLAGS"
        AC_EGREP_CPP(yes,[
-          #include <nss.h>
+	  #include <nss3/nss.h>
           #if NSS_VMAJOR >= 3 && NSS_VMINOR >= 2
              yes
           #endif
